Receiving ClosedChannelException instead of actual error

We aren't reporting a helpful error in very simple connectivity cases, like the server port is closed. Here is what it looks like when using the interop client against a non-existent server: 

```
$ ./build/install/grpc-interop-testing/bin/test-client
Running test empty_unary
...
Exception in thread "main" io.grpc.StatusRuntimeException: UNAVAILABLE
        at io.grpc.Status.asRuntimeException(Status.java:431)
        at io.grpc.stub.ClientCalls.getUnchecked(ClientCalls.java:157)
        at io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:106)
        at io.grpc.testing.integration.TestServiceGrpc$TestServiceBlockingStub.emptyCall(TestServiceGrpc.java:212)
        at io.grpc.testing.integration.AbstractTransportTest.emptyUnary(AbstractTransportTest.java:164)
        at io.grpc.testing.integration.TestServiceClient.runTest(TestServiceClient.java:220)
        at io.grpc.testing.integration.TestServiceClient.run(TestServiceClient.java:209)
        at io.grpc.testing.integration.TestServiceClient.main(TestServiceClient.java:80)
Caused by: java.nio.channels.ClosedChannelException
```

It seems the problem may also impact other, much harder to diagnose, issues like client certificate validation (like in #1327).

I saw a flake of this with #1594
```
io.grpc.netty.NettyTransportTest > ping_afterTermination FAILED
    java.lang.AssertionError: expected:<Status{code=UNAVAILABLE, description=null, cause=null}> but was:<Status{code=UNKNOWN, description=null, cause=java.nio.channels.ClosedChannelException}>
```

With current master I get

```
Jakobs-MBP-4:interop-testing buchgr$ ./build/install/grpc-interop-testing/bin/test-client --use_tls=false
Jun 12, 2016 5:03:32 PM io.grpc.internal.ManagedChannelImpl <init>
INFO: [ManagedChannelImpl@239963d8] Created with target directaddress:///localhost/127.0.0.1:8080
Running test empty_unary
Jun 12, 2016 5:03:32 PM io.grpc.internal.ManagedChannelImpl maybeTerminateChannel
INFO: [ManagedChannelImpl@239963d8] Terminated
Exception in thread "main" io.grpc.StatusRuntimeException: UNAVAILABLE: Terminated fail fast stream.
	at io.grpc.Status.asRuntimeException(Status.java:534)
	at io.grpc.stub.ClientCalls.getUnchecked(ClientCalls.java:208)
	at io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:141)
	at io.grpc.testing.integration.TestServiceGrpc$TestServiceBlockingStub.emptyCall(TestServiceGrpc.java:356)
	at io.grpc.testing.integration.AbstractInteropTest.emptyUnary(AbstractInteropTest.java:174)
	at io.grpc.testing.integration.TestServiceClient.runTest(TestServiceClient.java:210)
	at io.grpc.testing.integration.TestServiceClient.run(TestServiceClient.java:198)
	at io.grpc.testing.integration.TestServiceClient.main(TestServiceClient.java:83)
Caused by: io.grpc.StatusException: INTERNAL: Connection closed with unknown cause
	at io.grpc.Status.asException(Status.java:551)
	at io.grpc.internal.DelayedClientTransport$3.run(DelayedClientTransport.java:327)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Shutting down
```

```
Jakobs-MBP-4:interop-testing buchgr$ ./build/install/grpc-interop-testing/bin/test-client
Jun 12, 2016 5:04:55 PM io.grpc.internal.ManagedChannelImpl <init>
INFO: [ManagedChannelImpl@239963d8] Created with target directaddress:///localhost/127.0.0.1:8080
Running test empty_unary
Jun 12, 2016 5:04:55 PM io.grpc.internal.ManagedChannelImpl maybeTerminateChannel
INFO: [ManagedChannelImpl@239963d8] Terminated
Exception in thread "main" io.grpc.StatusRuntimeException: UNAVAILABLE: Terminated fail fast stream.
	at io.grpc.Status.asRuntimeException(Status.java:534)
	at io.grpc.stub.ClientCalls.getUnchecked(ClientCalls.java:208)
	at io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:141)
	at io.grpc.testing.integration.TestServiceGrpc$TestServiceBlockingStub.emptyCall(TestServiceGrpc.java:356)
	at io.grpc.testing.integration.AbstractInteropTest.emptyUnary(AbstractInteropTest.java:174)
	at io.grpc.testing.integration.TestServiceClient.runTest(TestServiceClient.java:210)
	at io.grpc.testing.integration.TestServiceClient.run(TestServiceClient.java:198)
	at io.grpc.testing.integration.TestServiceClient.main(TestServiceClient.java:83)
Caused by: io.grpc.StatusException: UNAVAILABLE
	at io.grpc.Status.asException(Status.java:551)
	at io.grpc.internal.DelayedClientTransport$3.run(DelayedClientTransport.java:327)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
       ...
Caused by: javax.net.ssl.SSLHandshakeException: General OpenSslEngine problem
	at io.netty.handler.ssl.OpenSslContext$AbstractCertificateVerifier.verify(OpenSslContext.java:466)
	at org.apache.tomcat.jni.SSL.readFromSSL(Native Method)
	at io.netty.handler.ssl.OpenSslEngine.readPlaintextData(OpenSslEngine.java:365)
        ...
Caused by: java.security.cert.CertificateException: No subject alternative DNS name matching localhost found.
	at sun.security.util.HostnameChecker.matchDNS(HostnameChecker.java:204)
	at sun.security.util.HostnameChecker.match(HostnameChecker.java:95)
	at sun.security.ssl.X509TrustManagerImpl.checkIdentity(X509TrustManagerImpl.java:455)
        ...
Shutting down
```

I just saw a flake at https://travis-ci.org/grpc/grpc-java/jobs/140586965 (since rebuilt):
```
io.grpc.netty.NettyTransportTest > ping_afterTermination FAILED
    java.lang.AssertionError: expected:<Status{code=UNAVAILABLE, description=null, cause=null}> but was:<Status{code=UNKNOWN, description=null, cause=java.nio.channels.ClosedChannelException}>
```

Flaked at https://travis-ci.org/grpc/grpc-java/jobs/140953978 (since rebuilt):
```
 io.grpc.netty.NettyTransportTest > ping_afterTermination FAILED
     java.lang.AssertionError: expected:<Status{code=UNAVAILABLE, description=null, cause=null}> but was:<Status{code=UNKNOWN, description=null, cause=java.nio.channels.ClosedChannelException}>
```

Seen again at https://travis-ci.org/grpc/grpc-java/jobs/143414728 (before the additional logging info) (since rebuilt)
```
io.grpc.netty.NettyTransportTest > ping_afterTermination FAILED
    java.lang.AssertionError: expected:<Status{code=UNAVAILABLE, description=null, cause=null}> but was:<Status{code=UNKNOWN, description=null, cause=java.nio.channels.ClosedChannelException}>
```

Got one with logging enabled

```java
io.grpc.netty.NettyTransportTest > ping_afterTermination FAILED
    java.lang.AssertionError: java.nio.channels.ClosedChannelException
    	at io.netty.channel.AbstractChannel$AbstractUnsafe.write(...)(Unknown Source)
     expected:<Status{code=UNAVAILABLE, description=null, cause=null}> but was:<Status{code=UNKNOWN, description=null, cause=java.nio.channels.ClosedChannelException}>
```

Will look next week.

--- a/core/src/main/java/io/grpc/internal/Http2Ping.java
+++ b/core/src/main/java/io/grpc/internal/Http2Ping.java
@@ -174,7 +174,7 @@ public class Http2Ping {
       this.callbacks = null;
     }
     for (Map.Entry<ClientTransport.PingCallback, Executor> entry : callbacks.entrySet()) {
-      doExecute(entry.getValue(), asRunnable(entry.getKey(), failureCause));
+      notifyFailed(entry.getKey(), entry.getValue(), failureCause);
     }
   }
 

--- a/netty/src/main/java/io/grpc/netty/NettyClientHandler.java
+++ b/netty/src/main/java/io/grpc/netty/NettyClientHandler.java
@@ -79,6 +79,7 @@ import io.netty.handler.codec.http2.Http2StreamVisitor;
 import io.netty.handler.codec.http2.StreamBufferingEncoder;
 import io.netty.handler.logging.LogLevel;
 
+import java.nio.channels.ClosedChannelException;
 import java.util.Random;
 import java.util.concurrent.Executor;
 import java.util.logging.Level;
@@ -472,7 +473,11 @@ class NettyClientHandler extends AbstractNettyHandler {
       @Override
       public void operationComplete(ChannelFuture future) throws Exception {
         if (!future.isSuccess()) {
-          finalPing.failed(future.cause());
+          Throwable cause = future.cause();
+          if (cause instanceof ClosedChannelException) {
+            cause = lifecycleManager.getShutdownThrowable();
+          }
+          finalPing.failed(cause);
           if (ping == finalPing) {
             ping = null;
           }

--- a/netty/src/main/java/io/grpc/netty/NettyClientHandler.java
+++ b/netty/src/main/java/io/grpc/netty/NettyClientHandler.java
@@ -476,6 +476,10 @@ class NettyClientHandler extends AbstractNettyHandler {
           Throwable cause = future.cause();
           if (cause instanceof ClosedChannelException) {
             cause = lifecycleManager.getShutdownThrowable();
+            if (cause == null) {
+              cause = Status.UNKNOWN.withDescription("Ping failed but for unknown reason.")
+                  .withCause(future.cause()).asException();
+            }
           }
           finalPing.failed(cause);
           if (ping == finalPing) {

--- a/netty/src/main/java/io/grpc/netty/NettyClientTransport.java
+++ b/netty/src/main/java/io/grpc/netty/NettyClientTransport.java
@@ -99,6 +99,8 @@ class NettyClientTransport implements ConnectionClientTransport {
 
   @Override
   public void ping(final PingCallback callback, final Executor executor) {
+    // The promise and listener always succeed in NettyClientHandler. So this listener handles the
+    // error case, when the channel is closed and the NettyClientHandler no longer in the pipeline.
     ChannelFutureListener failureListener = new ChannelFutureListener() {
       @Override
       public void operationComplete(ChannelFuture future) throws Exception {
@@ -238,13 +240,11 @@ class NettyClientTransport implements ConnectionClientTransport {
   private Status statusFromFailedFuture(ChannelFuture f) {
     Throwable t = f.cause();
     if (t instanceof ClosedChannelException) {
-      synchronized (this) {
-        Status shutdownStatus = lifecycleManager.getShutdownStatus();
-        if (shutdownStatus == null) {
-          return Status.UNKNOWN.withDescription("Channel closed but for unknown reason");
-        }
-        return shutdownStatus;
+      Status shutdownStatus = lifecycleManager.getShutdownStatus();
+      if (shutdownStatus == null) {
+        return Status.UNKNOWN.withDescription("Channel closed but for unknown reason");
       }
+      return shutdownStatus;
     }
     return Utils.statusFromThrowable(t);
   }

--- a/netty/src/main/java/io/grpc/netty/NettyClientTransport.java
+++ b/netty/src/main/java/io/grpc/netty/NettyClientTransport.java
@@ -236,6 +236,8 @@ class NettyClientTransport implements ConnectionClientTransport {
    * Convert ChannelFuture.cause() to a Status, taking into account that all handlers are removed
    * from the pipeline when the channel is closed. Since handlers are removed, you may get an
    * unhelpful exception like ClosedChannelException.
+   *
+   * <p>This method must only be called on the event loop.
    */
   private Status statusFromFailedFuture(ChannelFuture f) {
     Throwable t = f.cause();

--- a/netty/src/main/java/io/grpc/netty/Utils.java
+++ b/netty/src/main/java/io/grpc/netty/Utils.java
@@ -158,10 +158,7 @@ class Utils {
     if (s.getCode() != Status.Code.UNKNOWN) {
       return s;
     }
-    // TODO(ejona): reenable once startup races are resolved; ClosedChannelException is being seen
-    // still. Some tests are asserting UNAVAILABLE and were "working" previously but are now
-    // detecting that our behavior is flaky. See #1330
-    if (false && t instanceof ClosedChannelException) {
+    if (t instanceof ClosedChannelException) {
       // ClosedChannelException is used any time the Netty channel is closed. Proper error
       // processing requires remembering the error that occurred before this one and using it
       // instead.

--- a/testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java
+++ b/testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java
@@ -52,7 +52,6 @@ import static org.mockito.Mockito.reset;
 import static org.mockito.Mockito.timeout;
 import static org.mockito.Mockito.verify;
 
-import com.google.common.base.Throwables;
 import com.google.common.collect.Lists;
 import com.google.common.io.ByteStreams;
 import com.google.common.util.concurrent.MoreExecutors;
@@ -459,13 +458,7 @@ public abstract class AbstractTransportTest {
     }
     verify(mockPingCallback, timeout(TIMEOUT_MS)).onFailure(throwableCaptor.capture());
     Status status = Status.fromThrowable(throwableCaptor.getValue());
-    // TODO(buchgr): Remove once https://github.com/grpc/grpc-java/issues/1330 is resolved.
-    String stackTrace = "";
-    if (Status.UNAVAILABLE.getCode() != status.getCode()
-        && status.getCause() != null) {
-      stackTrace = Throwables.getStackTraceAsString(status.getCause());
-    }
-    assertCodeEquals(stackTrace, Status.UNAVAILABLE, status);
+    assertCodeEquals(Status.UNAVAILABLE, status);
   }
 
   @Test

